core/shared/src/main/scala/zio/metrics/Metric.scala

```diff
@@ -310,7 +325,7 @@ trait Metric[+Type, -In, +Out] extends ZIOAspect[Nothing, Any, Nothing, Any, Not
    * provided amount.
    */
   final def update(in: => In)(implicit trace: Trace): UIO[Unit] =
-    ZIO.succeed(unsafe.update(in, Set.empty)(Unsafe.unsafe))
+    ZIO.succeed(unsafe.update(in, Set.empty)(Unsafe.unsafe)) *> notifyListeners(in)
```

Comment:
This is much more expensive than the old version. I think what you should do is inside of `ZIO.succeed`, impurely perform the notification of listeners, so we can avoid the overhead. 

Of course, this assumes all listeners are unsafe.

Reply:
Sure. What I ended up doing is that `Unsafe.update` handles the impure notification of listeners.